src/experiments/experiment.js

```diff
@@ -0,0 +1,17 @@
+const { addFilter } = wp.hooks;
+
+/**
+ * Register a new experiment type.
+ *
+ * @param {object} experiment Experiement options.
+ * @param {int} experiment.id Experiement test ID.
+ * @param {string} experiment.title Experiement test title.
+ * @param {string} experiment.singleTitle Experiement test single entry title.
+ * @param {React.ReactNode} experiment.component Experiement test panel field component.
+ * @param {object} experiment.dispatcher Experiement test panel dispatch handler.
+ * @param {object} experiment.selector Experiement test panel select handler.
+ */
+export const registerExperiment = experiment => {
+	// Register the panel for the experiement.
+	addFilter( 'altis.experiments.experiments', `altis.experiments.${ experiment.id }`, experiments => [ ...experiments, experiment ] );
```

Comment:
I think as this follows the pattern of having to register editor blocks on both the backend and frontend maybe we can use the `block.json` pattern 🤔 I digress anyway, that's more for implementors to decide.

Anyway - I think what would be beneficial here is to validate that the experiment being registered is also registered on the backend, otherwise it's not going to do anything and the UI would just result in a ton of API errors.

Maybe `registerExperiment()` can follow the same conventions as `registerBlockType()`, `register_block_type()` and `register_post_ab_test()` e.g. the arguments are `func( id/name, settings/options )`.

Having the test ID as the first argument makes it easier to validate.

Reply:
Hmm, things is that the experiment ID is passed down with the experiment object down to all the components, otherwise I'll pass down the ID separately, OR merge the ID back into the object within the filter call here.

```js
/**
 * Register a new experiment type.
 *
 * @param {int} id Experiement test ID.
 * @param {object} options Experiement options.
 * @param {string} options.title Experiement test title.
 * @param {string} options.singleTitle Experiement test single entry title.
 * @param {React.ReactNode} options.component Experiement test panel field component.
 * @param {object} options.dispatcher Experiement test panel dispatch handler.
 * @param {object} options.selector Experiement test panel select handler.
 */
export const registerExperiment = ( id, options ) => {
	// Register the panel for the experiement.
	addFilter( 'altis.experiments.experiments', `altis.experiments.${ id }`, experiments => ( [
		...experiments,
		{
			id,
			...options,
		},
	] ) );
};
```

How would the validation work in this case ? Within the editor plugin code ? And how would passing the ID separately facilitate it ?

Reply:
Yeah you could merge it in to the object or have an object that acts as a registry, it doesn’t necessarily need to use a filter here looking at it. I think I suggested a filter thinking we’d have a default control and you just filter the variant input control.

Regarding validation you could serialise the registered AB tests when enqueueing the framework and cross reference that. I was expecting to share some data from the server side like the title / label etc

Reply:
Hmmm, I think with the structure and generic fields we have, we can probably register stuff in PHP, pass that to the editor, and use the frontend to only _override_ default options for a specific experiment, that way we take care of validation, and minimize the duplication between PHP and JS code.

We can then pass down the experiment ID only, and grab the experiment object from the registry without having to pass down the full object everywhere.

re filters, I had a bunch of filters earlier, but they trimmed down to one, as I now pass the experiment option itself, so I thought the same about using a registry object, probably facilitate by the PHP<>JS communication just mentioned above.

Reply:
Latest refactor makes the JS hooks completely optional, as you can create an experiment with an arbitrary value, and use that on the front-end as you please. We just need to document this, and the APIs that can be used for custom handling on front-end.

I've removed all hooks, and only used them to override test settings, which will be the same API we'll provide to custo integrations.